### PR TITLE
doc: add shared memory based inter-vm communication introduction

### PR DESCRIPTION
**Pull request by liuyuan-GitHub**
The shared memory based inter-vm communication can support high
performance communication between VMs including pre-launched and
post-launched VMs.

There are two solutions. One is DM-land which emulates a virtual
PCI device (called ivshmem device) to expose a shared memory region
base address and size and the shared memory region is allocated from
service OS.Another is Hypervisor-land which emulates ivshmem device
in the hypervisor and shared memory is allocated from hypervisor.

Signed-off-by: Yuan Liu <yuan1.liu@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/developer-guides/hld/hld-emulated-devices.rst [hunk 1]
```diff
@@ -22,3 +22,4 @@ documented in this section.
    Hostbridge emulation <hostbridge-virt-hld>
    AT keyboard controller emulation <atkbdc-virt-hld>
    Split Device Model <split-dm>
+   Inter-VM communication based on shared memory <ivshmem-hld>
```
Comment (ywan170) [inline doc/developer-guides/hld/hld-emulated-devices.rst:25]: Shared memory based inter-vm communication

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
+.. _ivshmem-hld:
+
+ACRN Shared Memory Based Inter-VM Communication
+###############################################
+
+To support inter virtual machines communication based on shared memory
+mechanism, ACRN device model/hypervisor emulate a virtual PCI device
+(Called ivshmem device) to expose the shared memory base address and
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:8]: to expose the base address and size of shared memory.

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+size. Currently, ACRN only supports post-launched to post-launched
+communication via ivshmem device. In future, ACRN will supports
+pre-launched VM inter-VM communication.
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:11]: Remove "Currently, ACRN only supports post-launched to post-launched communication via ivshmem device. In future, ACRN will supports pre-launched VM inter-VM communication."
We can talk it later, not at the beginning.

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+Inter-VM Communication Overview
+*******************************
+
+.. figure:: images/ivshmem-image1.png
+   :align: center
+   :name: ivshmem-architecture-overview
+
+   ACRN Inter-VM Communication Architecture Based on Shared Memory
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:20]: ACRN shared memory based inter-vm communication architecture

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+Two solutions for ivshmem device implementation, one is called DM-land,
+another is called Hypervisor-land. For DM-land, the ivshmem device is
+emulated in device model and the inter-VM communication shared memory
+region is allocated from Service memory. For Hypervisor-land, the
+ivshmem device is emulated in hypervisor and the shared memory is
+allocated from hypervisor. The former only work for post-launched VMs,
+the latter can work for both of pre-launched and post-launched VMs. The
+Hypervisor-land will be implemented in the future.
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:29]: Two solutions of ivshmem device: DM-land and hypervisor-land.  For DM-land, the ivshmem device emulated in acrn device model and its shared memory region is allocated from SOS's memory space. For hypervisor-land, it emulated in hypervisor and the shared memory regions are reserved in hypervisor's memory space. The former solution only support interactive between port-launched VMs and the latter one can work for both pre-launched and post-launched VMs. The hypervisor-land solution is WIP, plan to support in the near future.

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+-  Ivshmem dm – The ivshmem device model implements device related
+   registers and shared memory mapping. In the future it will also
+   support the interrupt mechanism.
+
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:34]: The ivshmem device model implements register virtualization and shared memory mapping. Will support notification/interrupt mechanism in future.

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+-  Ivshmem server – A daemon for inter-VM notification capability,
+   currently, it has not been implemented, so the inter-VM communication
+   doesn’t support notification mechanism.
+
+Ivshmem Device Introduction
+***************************
+
+The ivshmem device is a virtual standard PCI device, it consists of two
+BARs(BAR0 and BAR2), one is used for interrupt related register and BAR
+size is 256 bytes, another is used for exposing shared memory base
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:44]: one is used for emulate interrupt related registers, another one is used for exposing shared memory region. The ivshmem device doesn't support any extra capabilities.

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+Inter-VM Communication Overview
+*******************************
+
+.. figure:: images/ivshmem-image1.png
+   :align: center
+   :name: ivshmem-architecture-overview
+
+   ACRN shared memory based inter-vm communication architecture
+
+Two solutions of ivshmem device: DM-land and hypervisor-land. For DM-land,
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:20]: DM-land -> dm-land

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+the ivshmem device emulated in acrn device model and its shared memory region
+is allocated from SOS's memory space. For hypervisor-land, it emulated in
+hypervisor and the shared memory regions are reserved in hypervisor's memory
+space. The former solution only support interactive between port-launched VMs
```
Comment (ywan170) [inline doc/developer-guides/hld/ivshmem-hld.rst:24]: support -> supports

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by ywan170** [comment]
@NanlinXie Reviewed done, please go ahead.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by wenlingz** [comment]
@liuyuan-GitHub please refer make doc fail log:
make -C /root/workspace/acrn-hypervisor-premerge-build-image/acrn-hypervisor/doc html BUILDDIR=/root/workspace/acrn-hypervisor-premerge-build-image/acrn-hypervisor/build/doc
make[1]: Entering directory '/root/workspace/acrn-hypervisor-premerge-build-image/acrn-hypervisor/doc'

New errors/warnings found, please fix them:
==============================================

/root/workspace/acrn-hypervisor-premerge-build-image/acrn-hypervisor/build/doc/rst/developer-guides/hld/ivshmem-hld.rst:18: WARNING: image file not readable: developer-guides/hld/images/ivshmem-image1.png

make[1]: Leaving directory '/root/workspace/acrn-hypervisor-premerge-build-image/acrn-hypervisor/doc'
make[1]: *** [Makefile:61: html] Error 1
make: *** [Makefile:207: doc] Error 2

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 10]
```diff
@@ -0,0 +1,93 @@
+.. _ivshmem-hld:
+
+ACRN Shared Memory Based Inter-VM Communication
+###############################################
+
+To support inter virtual machines communication based on shared memory
+mechanism, ACRN device model/hypervisor emulate a virtual PCI device
+(Called ivshmem device) to expose the base address and size of shared
```
Comment (gvancuts) [inline doc/developer-guides/hld/ivshmem-hld.rst:8]:
```suggestion
(called ``ivshmem`` device) to expose the base address and size of shared
```

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+.. figure:: images/ivshmem-image1.png
+   :align: center
+   :name: ivshmem-architecture-overview
+
+   ACRN shared memory based inter-vm communication architecture
+
+Two solutions of ivshmem device: dm-land and hypervisor-land. For dm-land,
+the ivshmem device emulated in acrn device model and its shared memory region
+is allocated from SOS's memory space. For hypervisor-land, it emulated in
```
Comment (gvancuts) [inline doc/developer-guides/hld/ivshmem-hld.rst:22]:
```suggestion
is allocated from the Service VM's memory space. For hypervisor-land, it is emulated in
```

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+Inter-VM Communication Overview
+*******************************
+
+.. figure:: images/ivshmem-image1.png
+   :align: center
+   :name: ivshmem-architecture-overview
+
+   ACRN shared memory based inter-vm communication architecture
+
+Two solutions of ivshmem device: dm-land and hypervisor-land. For dm-land,
```
Comment (gvancuts) [inline doc/developer-guides/hld/ivshmem-hld.rst:20]: Are these two solutions two components of a single solution, or are these alternatives to each other?

I'm not a big fan of calling these `dm-land` and `hyervisor-land`, we've not used these in the past. I think I would prefer to use plain English to say if something runs in the Service VM or in the Hypervisor. @dbkinder , thoughts?

File: doc/developer-guides/hld/ivshmem-hld.rst [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+the ivshmem device emulated in acrn device model and its shared memory region
+is allocated from SOS's memory space. For hypervisor-land, it emulated in
+hypervisor and the shared memory regions are reserved in hypervisor's memory
+space. The former solution only supports interactive between port-launched VMs
```
Comment (gvancuts) [inline doc/developer-guides/hld/ivshmem-hld.rst:24]:
```suggestion
space. The former solution only supports interactive between post-launched VMs
```

File: doc/developer-guides/hld/ivshmem-hld.rst (hunk 12 above)
Comment (liuyuan-GitHub) [reply doc/developer-guides/hld/ivshmem-hld.rst:20]: > Are these two solutions two components of a single solution, or are these alternatives to each other?

Currently, they are alternatives. dm-land shared memory is allocated from Service VM, hypervisor-land shared memory is allocated from hypervisor.  In the future, when the hypervisor-land is implemented, we will check whether these two solutions can be merged into one solution.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: doc/developer-guides/hld/ivshmem-hld.rst (hunk 12 above)
Comment (gvancuts) [reply doc/developer-guides/hld/ivshmem-hld.rst:20]: > > Are these two solutions two components of a single solution, or are these alternatives to each other?
> 
> Currently, they are alternatives. dm-land shared memory is allocated from Service VM, hypervisor-land shared memory is allocated from hypervisor. In the future, when the hypervisor-land is implemented, we will check whether these two solutions can be merged into one solution.

Thanks for the clarification!

**Pull request merged**